pkg/querier/stats/stats.go

```diff
@@ -52,194 +58,213 @@ func (s *Stats) AddWallTime(t time.Duration) {
 
 	atomic.AddUint64(&s.FetchedIndexBytes, indexBytes)
 }
 
-func (s *Stats) LoadFetchedIndexBytes() uint64 {
+func (s *QueryStats) LoadFetchedIndexBytes() uint64 {
 	if s == nil {
 		return 0
 	}
 
 	return atomic.LoadUint64(&s.FetchedIndexBytes)
 }
 
-func (s *Stats) AddShardedQueries(num uint32) {
+func (s *QueryStats) AddShardedQueries(num uint32) {
 	if s == nil {
 		return
 	}
 
 	atomic.AddUint32(&s.ShardedQueries, num)
 }
 
-func (s *Stats) LoadShardedQueries() uint32 {
+func (s *QueryStats) LoadShardedQueries() uint32 {
 	if s == nil {
 		return 0
 	}
 
 	return atomic.LoadUint32(&s.ShardedQueries)
 }
 
-func (s *Stats) AddSplitQueries(num uint32) {
+func (s *QueryStats) AddSplitQueries(num uint32) {
 	if s == nil {
 		return
 	}
 
 	atomic.AddUint32(&s.SplitQueries, num)
 }
 
-func (s *Stats) LoadSplitQueries() uint32 {
+func (s *QueryStats) LoadSplitQueries() uint32 {
 	if s == nil {
 		return 0
 	}
 
 	return atomic.LoadUint32(&s.SplitQueries)
 }
 
-func (s *Stats) AddEstimatedSeriesCount(c uint64) {
+func (s *QueryStats) AddEstimatedSeriesCount(c uint64) {
 	if s == nil {
 		return
 	}
 
 	atomic.AddUint64(&s.EstimatedSeriesCount, c)
 }
 
-func (s *Stats) LoadEstimatedSeriesCount() uint64 {
+func (s *QueryStats) LoadEstimatedSeriesCount() uint64 {
 	if s == nil {
 		return 0
 	}
 
 	return atomic.LoadUint64(&s.EstimatedSeriesCount)
 }
 
-func (s *Stats) AddQueueTime(t time.Duration) {
+func (s *QueryStats) AddQueueTime(t time.Duration) {
 	if s == nil {
 		return
 	}
 
 	atomic.AddInt64((*int64)(&s.QueueTime), int64(t))
 }
 
-func (s *Stats) LoadQueueTime() time.Duration {
+func (s *QueryStats) LoadQueueTime() time.Duration {
 	if s == nil {
 		return 0
 	}
 
 	return time.Duration(atomic.LoadInt64((*int64)(&s.QueueTime)))
 }
 
-func (s *Stats) AddEncodeTime(t time.Duration) {
+func (s *QueryStats) AddEncodeTime(t time.Duration) {
 	if s == nil {
 		return
 	}
 
 	atomic.AddInt64((*int64)(&s.EncodeTime), int64(t))
 }
 
-func (s *Stats) LoadEncodeTime() time.Duration {
+func (s *QueryStats) LoadEncodeTime() time.Duration {
 	if s == nil {
 		return 0
 	}
 
 	return time.Duration(atomic.LoadInt64((*int64)(&s.EncodeTime)))
 }
 
-func (s *Stats) AddSamplesProcessed(c uint64) {
+func (s *QueryStats) AddSamplesProcessed(c uint64) {
 	if s == nil {
 		return
 	}
 
 	atomic.AddUint64(&s.SamplesProcessed, c)
 }
 
-func (s *Stats) LoadSamplesProcessed() uint64 {
+func (s *QueryStats) LoadSamplesProcessed() uint64 {
 	if s == nil {
 		return 0
 	}
 
 	return atomic.LoadUint64(&s.SamplesProcessed)
 }
 
-func (s *Stats) AddSpunOffSubqueries(num uint32) {
+func (s *QueryStats) AddSpunOffSubqueries(num uint32) {
 	if s == nil {
 		return
 	}
 	atomic.AddUint32(&s.SpunOffSubqueries, num)
 }
 
-func (s *Stats) LoadSpunOffSubqueries() uint32 {
+func (s *QueryStats) LoadSpunOffSubqueries() uint32 {
 	if s == nil {
 		return 0
 	}
 	return atomic.LoadUint32(&s.SpunOffSubqueries)
 }
 
+func (s *QueryStats) AddSamplesProcessedPerStep(points []StepStat) {
```

Comment:
shouldn't we be merging the points here? i think we share a `stats` instance for multiple queries, so the promql engine can invoke our `StatsRenderer` concurrently (that's why all fields are accessed atomically too i think)

Reply:
Makes sense